sdk/endpoint/endpoint.go

```diff
@@ -97,54 +111,71 @@ func (nsme *nsmEndpoint) Start() error {
 
-	span := spanhelper.FromContext(nsme.Context, fmt.Sprintf("Endpoint-%v-Start", nsme.Configuration.EndpointNetworkService))
-	span.LogObject("labels", nsme.Configuration.EndpointLabels)
+	nsme.registryClient = registry.NewNetworkServiceRegistryClient(nsme.GrpcClient)
+	for i := range nsme.registrations {
+		nsme.register(&nsme.registrations[i])
+	}
+
+	return nil
+}
+
+func (nsme *nsmEndpoint) register(r *registration) {
+	span := spanhelper.FromContext(nsme.Context, fmt.Sprintf("Endpoint-%v-Start", r.Name))
+	span.LogObject("labels", r.Labels)
 	defer span.Finish()
 
 	// Registering NSE API, it will listen for Connection requests from NSM and return information
 	// needed for NSE's forwarder programming.
 	nse := &registry.NetworkServiceEndpoint{
-		NetworkServiceName: nsme.Configuration.EndpointNetworkService,
+		NetworkServiceName: r.Name,
 		Payload:            "IP",
-		Labels:             tools.ParseKVStringToMap(nsme.Configuration.EndpointLabels, ",", "="),
+		Labels:             r.Labels,
 	}
 	registration := &registry.NSERegistration{
 		NetworkService: &registry.NetworkService{
-			Name:    nsme.Configuration.EndpointNetworkService,
+			Name:    r.Name,
 			Payload: "IP",
 		},
 		NetworkServiceEndpoint: nse,
 	}
 	span.LogObject("nse-request", registration)
 
-	nsme.registryClient = registry.NewNetworkServiceRegistryClient(nsme.GrpcClient)
 	registeredNSE, err := nsme.registryClient.RegisterNSE(span.Context(), registration)
 	if err != nil {
 		span.Logger().Fatalln("unable to register endpoint", err)
 	}
-	nsme.endpointName = registeredNSE.GetNetworkServiceEndpoint().GetName()
-	span.LogObject("endpoint-name", nsme.endpointName)
+	r.registeredName = registeredNSE.GetNetworkServiceEndpoint().GetName()
+	span.LogObject("endpoint-name", r.registeredName)
 	span.Logger().Infof("NSE registered: %v", registeredNSE)
 	span.Logger().Infof("NSE: channel has been successfully advertised, waiting for connection from NSM...")
-
-	return nil
 }
 
-func (nsme *nsmEndpoint) Delete() error {
-	span := spanhelper.FromContext(context.Background(), fmt.Sprintf("Endpoint-%v-Delete", nsme.Configuration.EndpointNetworkService))
+func (nsme *nsmEndpoint) deregister(r *registration) error {
+	span := spanhelper.FromContext(context.Background(), fmt.Sprintf("Endpoint-%v-Delete", r.Name))
 	defer span.Finish()
 	// prepare and defer removing of the advertised endpoint
 	removeNSE := &registry.RemoveNSERequest{
-		NetworkServiceEndpointName: nsme.endpointName,
+		NetworkServiceEndpointName: r.registeredName,
 	}
 	span.LogObject("delete-request", removeNSE)
 	_, err := nsme.registryClient.RemoveNSE(span.Context(), removeNSE)
 	if err != nil {
 		span.Logger().Errorf("Failed removing NSE: %v, with %v", removeNSE, err)
 	}
+	return err
+}
+
+func (nsme *nsmEndpoint) Delete() error {
+	var err error
+	for i := range nsme.registrations {
+		err1 := nsme.deregister(&nsme.registrations[i])
+		if err == nil && err1 != nil {
+			err = err1
+		}
```

Comment:
This check could ignore some errors...
Please consider using `errors.Wrapf` to put all errors into result.

Also, let's do not use names like `err1`. I would suggest `unregisterErr`

Reply:
Used `go-multierror` as in `client.go`.